django/core/mail/message.py

```diff
@@ -100,22 +100,62 @@ def forbid_multi_line_headers(name, val, encoding):
     return str(name), val
 
 
+def split_addr(addr, encoding):
+    """
+    Splits an address into local part and domain, properly encoded.
+    """
+    if '@' in addr:
+        localpart, domain = addr.split('@', 1)
+        localpart = Header(localpart, encoding).encode()
+        domain = domain.encode('idna').decode('ascii')
+    else:
+        localpart = Header(addr, encoding).encode()
+        domain = ''
+    return (localpart, domain)
+
+
 def sanitize_address(addr, encoding):
+    """
+    Formats a pair of (name, address) or an e-mail address string.
+
+    On Python 3, an `email.headerregistry.Address` object is used since
+    email.utils.formataddr naively encodes the name as ascii (see #25986).
+
+    On Python 2, fall back to the stdlib since `email.headerregistry` does not
+    exist.
+    """
     if not isinstance(addr, tuple):
         addr = parseaddr(force_text(addr))
     nm, addr = addr
+    localpart, domain = None, None
+
     nm = Header(nm, encoding).encode()
+
     try:
         addr.encode('ascii')
-    except UnicodeEncodeError:  # IDN
-        if '@' in addr:
-            localpart, domain = addr.split('@', 1)
-            localpart = str(Header(localpart, encoding))
-            domain = domain.encode('idna').decode('ascii')
+    except UnicodeEncodeError:  # IDN or non-ascii in the local part
+        localpart, domain = split_addr(addr, encoding)
+
+    # Do the feature test (Py2 vs. Py3)
```

Comment:
Can we use `six.PY2` so it's easy to identify what code can be removed when dropping Python 2?